x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java

```diff
@@ -316,11 +317,10 @@ private static Query translate(BinaryComparison bc, TranslatorHandler handler) {
                 return new RangeQuery(source, name, null, false, value, true, format, zoneId);
             }
             if (bc instanceof Equals || bc instanceof NullEquals || bc instanceof NotEquals) {
-                if (bc.left() instanceof FieldAttribute) {
```

Comment:
I've replaced the `instanceof` check with the assert above. To me it was quite confusing because it looked like `bc.left()` can also be of another type. But `handler.wrapFunctionQuery` in `doTranslate` ensures that `translate` is never called with an `BinaryComparison` with a `left()` that's not a `FieldAttribute`.

Reply:
Worth adding a comment or improving the return type of doTranslate.

Reply:
👍 I'm extending the interface of `doTranslate` to use the type system for handling this instead of assertions.